### PR TITLE
feat(core): add getUiState function

### PR DESCRIPTION
**Pull request by Haroenv**
**Summary**

This is useful for if you want to read the ui state externally from InstantSearch, like for example in https://github.com/algolia/autocomplete/pull/550

**Result**

a new `getUiState` function on the InstantSearch instance that can be used to retrieve the latest ui state, separately from the router.

### CODE REVIEW

**Comment by francoischalifour** [comment]
That'll be useful for external integrations to read the InstantSearch internal state! 🙂

File: src/lib/__tests__/InstantSearch-test.js [hunk 1]
```diff
@@ -101,6 +103,9 @@ See: https://www.algolia.com/doc/guides/building-search-ui/going-further/backend
   });
 
   it('throws if insightsClient is not a function', () => {
+    const warn = jest.spyOn(global.console, 'warn');
```
Comment (francoischalifour) [inline src/lib/__tests__/InstantSearch-test.js:106]: Why do we need these changes here?

File: src/lib/__tests__/InstantSearch-test.js [hunk 2]
```diff
@@ -2017,6 +2041,176 @@ See https://www.algolia.com/doc/guides/building-search-ui/widgets/customize-an-e
   });
 });
 
+describe('getUiState', () => {
+  test('retrieves empty ui state without widgets', () => {
```
Comment (francoischalifour) [inline src/lib/__tests__/InstantSearch-test.js:2045]:
```suggestion
  test('retrieves empty UI state without widgets', () => {
```

**Review by francoischalifour — APPROVED** [review]

File: src/lib/__tests__/InstantSearch-test.js (hunk 1 above)
Comment (Haroenv) [reply src/lib/__tests__/InstantSearch-test.js:106]: a continuation of #4746, the warnings now take up a lot of place, so I removed them from this file.

**Pull request merged**